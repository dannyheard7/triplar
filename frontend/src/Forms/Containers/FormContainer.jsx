import React from "react";
import {inputParsers} from "../utils/inputParsers.js";

export default class FormContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            errors: []
        };

        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        const data = this.parseFormData(event);

        this.props.apiFunction(data).then(({data}) => {
            let result = data.data.result;

            if (result.errors && result.errors.length > 0) {
                let errors = result.errors.reduce((dict, {field, messages}) => {
                    dict[field] = messages;
                    return dict;
                }, {});

                this.setState({errors: errors})
            } else {
                this.setState({errors: []});
                this.props.onSuccess(result);
            }
        }).catch(error => {
             console.log(error);
        });
    }

    parseFormData(event) {
        const form = event.target;
        const data = new FormData(event.target);

        for (let name of data.keys()) {
            const input = form.elements[name];
            const parserName = input.dataset.parse;

            if (parserName) {
                const parser = inputParsers[parserName];
                const parsedValue = parser(data.get(name));
                data.set(name, parsedValue);
            }
        }

        let variables = Object();
        for(let pair of data.entries()) {
           variables[pair[0]] = pair[1];
        }

        return variables;
    }

    render() {
        const props = {
            onSubmit: this.handleSubmit,
            errors: this.state.errors
        };

        return React.cloneElement(this.props.children, {...props});
    }

}