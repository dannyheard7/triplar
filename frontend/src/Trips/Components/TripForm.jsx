import React from "react";

import UncontrolledFormFieldGroup from "Forms/Components/UncontrolledFormFieldGroup";
import NonFieldErrors from "Forms/Components/NonFieldErrors";
import DateFieldGroup from "Forms/Components/DateFieldGroup";
import moment from "moment";


export default class  TripForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            startDate: null,
            endDate: null
        };

        this.onDateChange = this.onDateChange.bind(this);
        this.calculateEndDate = this.calculateEndDate.bind(this);
    }

    componentDidMount() {
        this.setState( {
            startDate: this.props.trip ? moment(this.props.trip.startDate) : moment(),
            endDate: this.props.trip ? moment(this.props.trip.endDate) : moment()
        });
    }

    componentDidUpdate(prevProps, prevState) {
        if(this.props.trip && (prevProps.trip.startDate !== this.props.trip.startDate
            || prevProps.trip.endDate !== this.props.trip.endDate)) {
            this.componentDidMount();
        }
    }

    onDateChange(value, date) {
        this.setState({
            [value]: moment(date, "DD/MM/YYYY")
        });
    }

    calculateEndDate(startDate) {
        const trip = this.props.trip ? this.props.trip : {};

        if(trip.endDate) {
            return this.state.endDate ? this.state.endDate : moment();
        } else {
            return startDate.clone().add(3, 'days');
        }
    }

    render() {
        const errors = this.props.errors;
        let trip = (this.props.trip) ? this.props.trip : {};

        let startDate = this.state.startDate ? moment(this.state.startDate) : moment();
        let endDate = this.calculateEndDate(startDate);

        return (
            <form onSubmit={this.props.onSubmit}>
                <NonFieldErrors errors={errors.non_field_errors}/>
                <UncontrolledFormFieldGroup errors={errors.name} label="Trip Name" name="name" type="text"
                                            value={trip.name} required={true}/>
                <DateFieldGroup errors={errors.startDate} label="Start Date" name="startDate"
                                onChange={(date) => this.onDateChange("startDate", date)}
                                value={startDate.format("DD/MM/YYYY")} required={true}/>
                <DateFieldGroup errors={errors.endDate} label="End Date" name="endDate"
                                onChange={(date) => this.onDateChange("endDate", date)}
                                value={endDate.format("DD/MM/YYYY")} required={true}/>
                <button type="submit" className="btn btn-primary">{this.props.submitLabel}</button>
            </form>
        );
    }
}

