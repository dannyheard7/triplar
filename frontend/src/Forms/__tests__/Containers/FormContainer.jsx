import React from "react";
import {shallow} from "enzyme";
import ShallowRenderer from 'react-test-renderer/shallow';
import TripForm from "Trips/Components/TripForm";
import FormContainer from "Forms/Containers/FormContainer";

const renderer = new ShallowRenderer();

const faker = require('faker');

const formData = FormData;

describe('<FormContainer />', () => {
    const event = {
        preventDefault: () => {
        },
        target: {}
    };

    const onSuccess = jest.fn();

    const apiFunc = jest.fn().mockImplementation(trip => {
        return new Promise((resolve, reject) => {
            let response = {status: 200, data: {data: {result: trip}}};
            resolve(response);
        });
    });

    const props = {apiFunction: apiFunc, onSuccess: onSuccess, children: <TripForm />};

    afterEach(() => {
        apiFunc.mockClear();
        onSuccess.mockClear();
    });

    test('renders correctly', () => {
        const result = renderer.render(<FormContainer {...props} />);
        expect(result).toMatchSnapshot();
    });

    test('calls handle submit after receiving submit from <FormContainer />', async() => {
        const spy = jest.spyOn(FormContainer.prototype, 'handleSubmit');

        let container = shallow(<FormContainer {...props}  />);
        container.instance().parseFormData = jest.fn().mockReturnValue({});

        container.find(TripForm).prop('onSubmit')(event);
        await Promise.resolve();

        expect(spy).toBeCalled();
    });

    test('calls apiFunction prop during handleSubmit()', async() => {
        let container = shallow(<FormContainer {...props}  />);
        container.instance().parseFormData = jest.fn().mockReturnValue({});

        container.instance().handleSubmit(event);
        await Promise.resolve();

        expect(apiFunc).toBeCalled();
    });

    test('calls onSuccess prop on api call success', async() => {
        let container = shallow(<FormContainer {...props}  />);
        container.instance().parseFormData = jest.fn().mockReturnValue({});

        container.find(TripForm).prop('onSubmit')(event);
        await Promise.resolve();

        expect(onSuccess).toBeCalled();
    });

    test('errors state change updates TripForm Prop', () => {
        const container = shallow(<FormContainer {...props} />);
        expect(container.find(TripForm).props().errors).toEqual(container.state('errors'));

        let errors = jest.fn();
        container.setState({errors: errors});

        expect(container.find(TripForm).props().errors).toEqual(errors);
    });

    test('unsuccesful api call updates state errors', async () => {
        const spy = jest.fn();
        const data = {result: {errors: [{field: faker.random.word(), messages: [faker.random.word()]}]}};
        const apiFunc = jest.fn().mockReturnValueOnce(new Promise((resolve, reject) => {
            let response = {status: 400, data: {data: data}};
            resolve(response);
        }));

        const container = shallow(<FormContainer apiFunction={apiFunc} onSuccess={onSuccess} children={TripForm}/>);
        container.instance().parseFormData = jest.fn().mockReturnValue({});

        container.instance().handleSubmit(event);
        await Promise.resolve();

        expect(spy).not.toBeCalled();
        expect(container.state('errors')).toEqual({[data.result.errors[0].field]: data.result.errors[0].messages});
    });

    test('onSuccess state.errors is set to empty object', async() => {
        let container = shallow(<FormContainer {...props}  />);
        container.instance().parseFormData = jest.fn().mockReturnValue({});

        container.find(TripForm).prop('onSubmit')(event);
        await Promise.resolve();

        expect(container.state('errors')).toEqual([]);
    });

    test.skip('parseFormData uses FormData to parse form', () =>{
        const event = {
            preventDefault: () => { },
            target: undefined
        };

        global.FormData = formData;

        let container = shallow(<FormContainer {...props}  />);

        container.instance().parseFormData(event);

        expect(formData).toBeCalled();
    });

});