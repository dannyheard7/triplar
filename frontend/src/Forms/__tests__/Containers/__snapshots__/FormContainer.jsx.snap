// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormContainer /> renders correctly 1`] = `
<TripForm
  errors={Array []}
  onSubmit={[Function]}
  trip={Object {}}
/>
`;
